### PR TITLE
Add a connection limiting listener

### PR DESCRIPTION
**Pull request by rosstimothy**
Introduces `limiter.Listener` to provide a consistent and resuable mechanism for limiting incoming connections per client. The new listener is used by `sshutils/server.go` instead of manually applying limits in `HandleConnection`.

This is particuarly important now that the Proxy SSH port multiplexes both SSH and gRPC. Each listener is now wrapped in a `limiter.Listener` that uses the same `limiter.ConnectionsListener` to ensure that the connection limits for the Proxy are enforced for all traffic on the port.

### CODE REVIEW

File: lib/limiter/connlimiter.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ import (
 // ConnectionsLimiter is a network connection limiter and tracker
 type ConnectionsLimiter struct {
 	*connlimit.ConnLimiter
-	*sync.Mutex
+	sync.Mutex
 	connections    map[string]int64
 	maxConnections int64
```
Comment (espadolini) [inline lib/limiter/connlimiter.go:34]:
```suggestion
	maxConnections int64

	mu          sync.Mutex
	connections map[string]int64
```

or something along those lines, to make it clear that the mutex is only protecting the map?

File: lib/limiter/connlimiter.go [hunk 2]
```diff
@@ -64,55 +62,62 @@ func (l *ConnectionsLimiter) WrapHandle(h http.Handler) {
...
+
 	if numberOfConnections >= l.maxConnections {
-		return trace.LimitExceeded(
-			"too many connections from %v: %v, max is %v",
-			token, numberOfConnections, l.maxConnections)
+		return trace.LimitExceeded("too many connections from %v: %v, max is %v", token, numberOfConnections, l.maxConnections)
 	}
-	l.connections[token] = numberOfConnections + 1
+
+	l.connections[token]++
```
Comment (espadolini) [inline lib/limiter/connlimiter.go:82]: This is a pessimization, technically? I hope the compiler is smart enough to see through it

File: lib/limiter/limiter.go [hunk 3]
```diff
@@ -150,7 +154,7 @@ func (l *Limiter) UnaryServerInterceptorWithCustomRate(customRate CustomRateFunc
 	}
 }
 
-// StreamServerInterceptor is a GPRC stream interceptor that rate limits
+// StreamServerInterceptor is a gPRC stream interceptor that rate limits
```
Comment (espadolini) [inline lib/limiter/limiter.go:157]:
```suggestion
// StreamServerInterceptor is a gRPC stream interceptor that rate limits
```

**Review by espadolini — APPROVED** [review]
It might be out of scope for this PR, but it could be nice to have `limiter.ConnectionLimiter` itself return a callable to release the connection, rather than having to do it by hand for the listener.

**Review by greedy52 — APPROVED** [review]

**Pull request merged**